### PR TITLE
Deprecate --nocpp command line flag

### PR DESCRIPTION
**Pull request by nurmukhametov**
This PR partially addresses #3150

### CODE REVIEW

**Review by dbabokin — COMMENTED** [review]
Why do we want to remove this flag? Why not fixing it?

**Comment by dbabokin** [comment]
> > Why do we want to remove this flag? Why not fixing it?
> 
> It contradicts the current design, which relies on internal headers.
> 
> I also assume this flag was supposed to be used in tests only, but it is not actually used a lot. That's why I think it is easier to just deprecate and eventually remove it rather than fix it.

My understanding is that you can manually run preprocessor separately and run the compiler separately. I.e. run `-E` and then `--nocpp`. It might be for debugging purposes, not for testing. I guess that's the main use case. If we can fix that, I think we should fix that just to look like C++ compiler from the prospective of using `-E`/`--nocpp`. The question is how easy to make it work with new stdlib logic. If it's really messy, we can remove it.

Though we should probably give more detailed explanation in the deprecation notice.

**Comment by nurmukhametov** [comment]
> My understanding is that you can manually run preprocessor separately and run the compiler separately. I.e. run `-E` and then `--nocpp`. It might be for debugging purposes, not for testing. I guess that's the main use case.

This use-case, i.e., something like this:

```bash
$ ispc -E test.ispc -o - | ispc --nocpp - --emit-llvm-text -o -
```

works with reasonably same ISPC compiler.

At the same time, some tests that can be compiled before PR #2743 are not compiled now (like the test in this PR).

**Comment by dbabokin** [comment]
> works with reasonably same ISPC compiler.

It should work for the same compiler in this use-case scenario. It's for the user to debug what compiler does with the source code and try to understand the problems in user code. It's not to debug compiler itself.

> At the same time, some tests that can be compiled before PR #2743 are not compiled now (like the test in this PR).

So, the question is how difficult to make it work as previously. If it's not straightforward, I'm ok with depreciation. But if we can make it work relatively easy, we should keep it.

**Comment by aneshlya** [comment]
> So, the question is how difficult to make it work as previously. If it's not straightforward, I'm ok with depreciation. But if we can make it work relatively easy, we should keep it.

It's not straightforward, that's why this PR is here in the first place.
Also this option is broken (like in the test in this PR) in the last release and no one complained :)

**Comment by dbabokin** [comment]
> It's not straightforward, that's why this PR is here in the first place. Also this option is broken (like in the test in this PR) in the last release and no one complained :)

Ok, then let's go with this PR.

**Review by dbabokin — APPROVED** [review]

File: docs/ispc.rst [hunk 1]
```diff
@@ -674,6 +674,8 @@ Updating ISPC Programs For Changes In ISPC 1.26.0
 Macro definitions for LLVM version that ISPC is based on were added. Please,
 refer to `The Preprocessor`_ for more details.
 
+``--nocpp`` command line flag is deprecated and will be removed in future.
+
```
Comment (aneshlya) [inline docs/ispc.rst:678]: There is a mention of `--nocpp` flag in Preprocessor section - I think it can be removed.

Comment (nurmukhametov) [reply docs/ispc.rst]: Done

Comment (aneshlya) [reply docs/ispc.rst]: I don't see ispc.rst in the modified files anymore.

Comment (nurmukhametov) [reply docs/ispc.rst:678]: Right, it has to be now.

**Review by aneshlya — APPROVED** [review]

**Pull request merged**